File: Opportunity.ChakraBridge.UWP/Browser/Console.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Text;
using System.Threading.Tasks;
using Windows.Foundation;

namespace Opportunity.ChakraBridge.UWP.Browser
{
    /// <summary>
    /// WinRT implementation of js Console object.
    /// </summary>
    public sealed class Console
    {
        private Console() { }

        /// <summary>
        /// Raises when any method of console has been invoked.
        /// </summary>
        public event TypedEventHandler<Console, ConsoleLoggingEventArgs> Logging;

        /// <summary>
        /// Get or create a console object, and put it to <see cref="JsValue.GlobalObject"/> of current context.
        /// </summary>
        /// <returns>
        /// The created console object.
        /// </returns>
        public static Console GetOrCreate()
        {
            try
            {
                if (JsValue.GlobalObject.Get("console") is JsObject oldCon &&
                    oldCon.Get("rtConsole") is JsObject jsStore &&
                    jsStore.ToInspectable() is Console c)
                    return c;
            }
            catch { }
            var con = new Console();
            var jsConStore = JsValue.Create(con);
            var callback = JsFunction.Create(OnLogging);
            var func = (JsFunction)JsContext.RunScript(@"
(function () {
    let logging = Symbol('LoggingCallback');
    let getStack = Symbol('GetStack');
    function Console(console, callback) {
        this.rtConsole = console;
        this[logging] = callback;
        Object.preventExtensions(this);
    }

    Console.prototype[logging] = null;
    Console.prototype[getStack] = function(){ try{ throw new Error();}catch(ex){return ex.stack;} }

    Console.prototype.assert = function (cond, ...data) { if (cond) this[logging]('assert', data, this[getStack]()); };
    Console.prototype.clear = function (...data) { this[logging]('clear', data, this[getStack]()); };
    Console.prototype.count = function (...data) { this[logging]('count', data, this[getStack]()); };
    Console.prototype.debug = function (...data) { this[logging]('debug', data, this[getStack]()); };
    Console.prototype.dir = function (...data) { this[logging]('dir', data, this[getStack]()); };
    Console.prototype.dirxml = function (...data) { this[logging]('dirxml', data, this[getStack]()); };
    Console.prototype.error = function (...data) { this[logging]('error', data, this[getStack]()); };
    Console.prototype.exception = Console.prototype.error;
    Console.prototype.group = function (...data) { this[logging]('group', data, this[getStack]()); };
    Console.prototype.groupCollapsed = function (...data) { this[logging]('groupCollapsed', data, this[getStack]()); };
    Console.prototype.groupEnd = function (...data) { this[logging]('groupEnd', data, this[getStack]()); };
    Console.prototype.info = function (...data) { this[logging]('info', data, this[getStack]()); };
    Console.prototype.log = function (...data) { this[logging]('log', data, this[getStack]()); };
    Console.prototype.profile = function (...data) { this[logging]('profile', data, this[getStack]()); };
    Console.prototype.profileEnd = function (...data) { this[logging]('profileEnd', data, this[getStack]()); };
    Console.prototype.table = function (...data) { this[logging]('table', data, this[getStack]()); };
    Console.prototype.time = function (...data) { this[logging]('time', data, this[getStack]()); };
    Console.prototype.timeEnd = function (...data) { this[logging]('timeEnd', data, this[getStack]()); };
    Console.prototype.timeStamp = function (...data) { this[logging]('timeStamp', data, this[getStack]()); };
    Console.prototype.trace = function (...data) { this[logging]('trace', data, this[getStack]()); };
    Console.prototype.warn = function (...data) { this[logging]('warn', data, this[getStack]()); };

    Object.preventExtensions(Console);
    Object.preventExtensions(Console.prototype);
    return Console;
})()

");
            var console = func.New(jsConStore, callback);
            JsValue.GlobalObject.Set("console", console);
            return con;
        }

        private static JsValue OnLogging(JsFunction callee, JsObject caller, bool isConstructCall, IList<JsValue> arguments)
        {
            var con = (Console)caller.Get("rtConsole").ToInspectable();
            var l = con.Logging;
            if (l is null)
                return null;
            l(con, new ConsoleLoggingEventArgs
            {
                Verb = arguments[0].ToString(),
                Data = (JsArray)arguments[1],
                CallStack = arguments[2].ToString().Substring("Error\n   at Console.prototype[getStack] (Unknown script code:12:52)\n   at Console.prototype.error (Unknown script code:20:52)\n".Length),
            });
            return null;
        }
    }

    /// <summary>
    /// Provide args for <see cref="Console.Logging"/> event.
    /// </summary>
    public sealed class ConsoleLoggingEventArgs
    {
        /// <summary>
        /// The name of function be invoked.
        /// </summary>
        public string Verb { get; internal set; }

        /// <summary>
        /// Arguments of invoke.
        /// </summary>
        public IReadOnlyList<JsValue> Data { get; internal set; }

        /// <summary>
        /// Js call stack of the invoke.
        /// </summary>
        public string CallStack { get; internal set; }
    }
}

File: Opportunity.ChakraBridge.WinRT/Browser/Console.cpp
#include "pch.h"
#include "Console.h"
#include <functional>

using namespace Opportunity::ChakraBridge::WinRT;
using namespace Opportunity::ChakraBridge::WinRT::Browser;

#define RT_EXT_CONSOLE_NAME L"__rt_external_console__"

RawValue Console::OnLogging(const RawValue& callee, const RawValue& caller, const bool isConstructCall, const RawValue*const arguments, const unsigned short argumentCount, const nullptr_t&)
{
    _ASSERTE(argumentCount == 3);
    auto con = reinterpret_cast<Console^>(caller[RT_EXT_CONSOLE_NAME]().ToInspectable());
    auto args = ref new ConsoleLoggingEventArgs();
    args->Verb = arguments[0].ToString();
    args->Data = (ref new JsArrayImpl(arguments[1]))->ArrayGetView();
    auto callstack = arguments[2].ToString();
    auto callstackptr = callstack.Data();
    auto lf = 0;
    while (*callstackptr != L'\0' && lf < 3)
    {
        if (*callstackptr == L'\n')
            lf++;
        callstackptr++;
    }
    args->CallStack = ref new string(callstackptr, unsigned int(callstack.Data() + callstack.Length() - callstackptr));

    con->Logging(con, args);
    return nullptr;
}

Console^ Console::GetOrCreate()
{
    try
    {
        auto oldConsole = dynamic_cast<JsObjectImpl^>(to_impl(JsValue::GlobalObject)->Lookup(L"console"));
        if (oldConsole != nullptr)
        {
            auto con = safe_cast<Console^>(to_impl(oldConsole)->Lookup(RT_EXT_CONSOLE_NAME)->ToInspectable());
            if (con != nullptr)
                return con;
        }
    }
    catch (...) {}
    auto con = ref new Console();
    auto jsCon = RawValue(reinterpret_cast<IInspectable*>(con));
    auto callback = RawValue::CreateFunction<nullptr_t, OnLogging>(nullptr);
    const auto func = RawContext::RunScript(LR"=(
    (function() {
        let logging = Symbol('LoggingCallback');
        let getStack = Symbol('GetStack');
        class Console
        {
            constructor(console, callback)
            {
                Object.defineProperty(this, ')=" RT_EXT_CONSOLE_NAME LR"=(', { get() { return console }, configurable: false, enumerable: false });
                Object.defineProperty(this, logging, { get() { return callback }, configurable: false, enumerable: false });
            }
            assert(cond, ...data)
            {
            if (cond)
                this[logging]('assert', data, this[getStack]());
            }
            clear(...data) { this[logging]('clear', data, this[getStack]()); }
            count(...data) { this[logging]('count', data, this[getStack]()); }
            debug(...data) { this[logging]('debug', data, this[getStack]()); }
            dir(...data) { this[logging]('dir', data, this[getStack]()); }
            dirxml(...data) { this[logging]('dirxml', data, this[getStack]()); }
            error(...data) { this[logging]('error', data, this[getStack]()); }
            exception(...data) { this[logging]('error', data, this[getStack]()); }
            group(...data) { this[logging]('group', data, this[getStack]()); }
            groupCollapsed(...data) { this[logging]('groupCollapsed', data, this[getStack]()); }
            groupEnd(...data) { this[logging]('groupEnd', data, this[getStack]()); }
            info(...data) { this[logging]('info', data, this[getStack]()); }
            log(...data) { this[logging]('log', data, this[getStack]()); }
            profile(...data) { this[logging]('profile', data, this[getStack]()); }
            profileEnd(...data) { this[logging]('profileEnd', data, this[getStack]()); }
            table(...data) { this[logging]('table', data, this[getStack]()); }
            time(...data) { this[logging]('time', data, this[getStack]()); }
            timeEnd(...data) { this[logging]('timeEnd', data, this[getStack]()); }
            timeStamp(...data) { this[logging]('timeStamp', data, this[getStack]()); }
            trace(...data) { this[logging]('trace', data, this[getStack]()); }
            warn(...data) { this[logging]('warn', data, this[getStack]()); }

            [getStack]() { try { throw new Error(); } catch (ex) { return ex.stack; } }
        }
        Object.preventExtensions(Console);
        Object.preventExtensions(Console.prototype);
        return Console;
    })()
)=", JS_SOURCE_CONTEXT_NONE, __FILEW__);
    auto console = func.New(jsCon, callback);
    RawValue::GlobalObject()[L"console"] = console;
    return con;
}
